include/openssl/lhash.h.in

```diff
@@ -201,18 +205,6 @@ void OPENSSL_LH_node_usage_stats_bio(const OPENSSL_LHASH *lh, BIO *out);
     { \
         return OPENSSL_LH_num_items((OPENSSL_LHASH *)lh); \
     } \
-    static ossl_unused ossl_inline void lh_##type##_node_stats_bio(const LHASH_OF(type) *lh, BIO *out) \
-    { \
-        OPENSSL_LH_node_stats_bio((const OPENSSL_LHASH *)lh, out); \
-    } \
-    static ossl_unused ossl_inline void lh_##type##_node_usage_stats_bio(const LHASH_OF(type) *lh, BIO *out) \
-    { \
-        OPENSSL_LH_node_usage_stats_bio((const OPENSSL_LHASH *)lh, out); \
-    } \
-    static ossl_unused ossl_inline void lh_##type##_stats_bio(const LHASH_OF(type) *lh, BIO *out) \
-    { \
-        OPENSSL_LH_stats_bio((const OPENSSL_LHASH *)lh, out); \
-    } \
```

Comment:
Leaving these in generates warnings due to calling deprecated functions. Any idea what to do about this?

Reply:
I'd say just keep the functions but remove the inner calls to the deprecated functions. However that also means we can just drop the bodies of the various OPENSSL_LH_*stats* functions in lh_stats.c because they won't be really used.

Effectively it won't be just deprecation but making all the LH_*stats* functions no-ops.

Reply:
Unfortunately that solution would be contrary to our vote in openssl/technical-policies#35

Reply:
A different solution: create a new macro `DEFINE_NO_STATS_LHASH_OF` (or something with a better name - imagination fails me), which supplies all of the same inline functions but missing out the deprecated ones. Use that internally everywhere we currently use `DEFINE_LHASH_OF`. Deprecate the old `DEFINE_LHASH_OF` macro.

Reply:
Hmm, but deprecation of DEFINE_LHASH_OF() is also outside of the OTC vote. Perhaps we should discuss this within OTC once more.

Reply:
My proposed solution seems like a natural extension of the existing vote...but ok, let's discuss at OTC.

Reply:
Technically, the obvious thing is to move these to another macro (see below) that simply gets used here:

``` diff
diff --git a/include/openssl/lhash.h.in b/include/openssl/lhash.h.in
index 198e5278e0..911563f800 100644
--- a/include/openssl/lhash.h.in
+++ b/include/openssl/lhash.h.in
@@ -169,6 +169,21 @@ OSSL_DEPRECATEDIN_3_1 void OPENSSL_LH_node_usage_stats_bio(const OPENSSL_LHASH *
     } \
     LHASH_OF(type)
 
+# ifndef OPENSSL_NO_DEPRECATED_3_1
+#  define DEFINE_LHASH_STAT_FNS_OF(type) \
+    OSSL_DEPRECATEDIN_3_1 static ossl_unused ossl_inline void \
+    lh_##type##_node_stats_bio(const LHASH_OF(type) *lh, BIO *out)      \
+    { \
+        OPENSSL_LH_node_stats_bio((const OPENSSL_LHASH *)lh, out); \
+    } \
+    OSSL_DEPRECATEDIN_3_1 static ossl_unused ossl_inline void \
+    lh_##type##_node_usage_stats_bio(const LHASH_OF(type) *lh, BIO *out) \
+    { \
+        OPENSSL_LH_node_usage_stats_bio((const OPENSSL_LHASH *)lh, out); \
+    }
+# else
+#  define DEFINE_LHASH_STAT_FNS_OF(type)
+# endif
 # define DEFINE_LHASH_OF(type) \
     LHASH_OF(type) { union lh_##type##_dummy { void* d1; unsigned long d2; int d3; } dummy; }; \
     static ossl_unused ossl_inline LHASH_OF(type) *lh_##type##_new(unsigned long (*hfn)(const type *), \
@@ -205,6 +220,7 @@ OSSL_DEPRECATEDIN_3_1 void OPENSSL_LH_node_usage_stats_bio(const OPENSSL_LHASH *
     { \
         return OPENSSL_LH_num_items((OPENSSL_LHASH *)lh); \
     } \
+    DEFINE_LHASH_STAT_FNS_OF(type) \
     static ossl_unused ossl_inline unsigned long lh_##type##_get_down_load(LHASH_OF(type) *lh) \
     { \
         return OPENSSL_LH_get_down_load((OPENSSL_LHASH *)lh); \
```

But, as was mentioned in the OTC discussion today, this generates a lot of warnings about use of deprecation functions.  The way we usually handle this, is to add this in the beginning of the affected `.c` files:

``` C
#define OPENSSL_SUPPRESS_DEPRECATED
```

However, I fear that's going to be a lot of files (no, I haven't made the effort), which gets us back the question I tried to raise when `OPENSSL_SUPPRESS_DEPRECATED` was invented: when do we get to a point when we might as well amend the `--strict-warnings` set of flags to simply ignore deprecation warnings?

Reply:
To allow deprecation of static functions:

``` diff
diff --git a/include/openssl/macros.h b/include/openssl/macros.h
index 42e35ed9d3..3f3bdbc444 100644
--- a/include/openssl/macros.h
+++ b/include/openssl/macros.h
@@ -76,8 +76,8 @@
  * are unsuitable for use in a typedef.
  */
 # ifndef OSSL_DEPRECATED
-#  define OSSL_DEPRECATED(since)                extern
-#  define OSSL_DEPRECATED_FOR(since, message)   extern
+#  define OSSL_DEPRECATED(since)
+#  define OSSL_DEPRECATED_FOR(since, message)
 # endif
 
 /*
```

Frankly, I can't recall why we defined them to `extern`, that seems utterly unnecessary...  I think it comes from an early iteration of that header file, and never got cleaned up.

Reply:
To turn off deprecation errors with `--strict-warnings`:

``` diff
diff --git a/Configure b/Configure
index 8b0b25d5f6..1321c18ae5 100755
--- a/Configure
+++ b/Configure
@@ -157,6 +157,7 @@ my @gcc_devteam_warn = qw(
     -Wextra
     -Wno-unused-parameter
     -Wno-missing-field-initializers
+    -Wno-deprecated-declarations
     -Wswitch
     -Wsign-compare
     -Wshadow
@@ -187,8 +188,10 @@ my @clang_devteam_warn = qw(
     -Wmissing-variable-declarations
 );
 
+# /wd4996 is the MSVC equivalent of -Wno-deprecated-declaration
 my @cl_devteam_warn = qw(
     /WX
+    /wd4996
 );
 
 my $strict_warnings = 0;
```

With that and my previous diffs, this PR builds cleanly.

Quite frankly, between spraying `#include "internal/deprecated.h"` all over the place or this, I prefer this.

Reply:
> Frankly, I can't recall why we defined them to `extern`, that seems utterly unnecessary... I think it comes from an early iteration of that header file, and never got cleaned up.

I think it caused some problems on some compilers or something. I do not remember the exact reason but there was some reason for that.

Reply:
The extern stops the compiler emitting a copy of the function no matter what.  Such a function is only used for inlining.  Even it it's address is taken, the local copy won't happen (this case would need an external definition too).